File: src/typescript/models/Crag.ts
import {
  Column,
  Entity,
  OneToMany,
  PrimaryGeneratedColumn,
  OneToOne,
  JoinColumn,
} from "typeorm";

import Coordinate, { CoordinateOptional } from "./Coordinate.js";
import Area from "./Area.js";
import { cascadeOneToMany } from "../db/cascadeOptions.js";
import Commentable from "./Commentable.js";
import Trail from "./Trail.js";
import Bounds from "./Bounds.js";

@Entity()
export default class Crag {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  name: string;

  @Column({ nullable: true })
  description?: string;

  @OneToOne((type) => Bounds, (bounds) => bounds.crag, {
    nullable: true,
    onDelete: "SET NULL",
  })
  bounds?: Bounds;

  @Column((type) => CoordinateOptional)
  _center?: CoordinateOptional;

  get center(): Coordinate | undefined {
    if (this._center.lat && this._center.lng) {
      return new Coordinate(this._center.lat, this._center.lng);
    }
  }

  @Column("int")
  defaultZoom: number;
  @Column("int")
  minZoom: number;
  @Column("int")
  maxZoom: number;

  // Relationships
  @OneToMany((type) => Area, (area) => area.crag, cascadeOneToMany)
  areas: Area[];

  @OneToOne((type) => Commentable, { nullable: true, onDelete: "SET NULL" })
  @JoinColumn()
  commentable?: Commentable;

  @OneToOne((type) => Trail, { nullable: true, onDelete: "SET NULL" })
  @JoinColumn()
  trail?: Trail;

  // @todo See fetchCragContainer todo
  _isLoaded?: boolean;

  toJSON(): any {
    return Object.keys(this).reduce(
      (memo, k) => ({
        ...memo,
        [k]: this[k],
      }),
      {
        center: this.center,
      }
    );
  }
}

File: clutter/lib/models/crag.dart
import 'area.dart';
import 'bounds.dart';
import 'coordinate.dart';
import 'trail.dart';
import 'types.dart';

class Crag {
  final int id;
  final String name;
  final String? description;
  final Bounds bounds;
  final Coordinate center;
  final int defaultZoom;
  final int minZoom;
  final int maxZoom;
  final List<Area> areas;
  final Trail? trail;

  const Crag({
    required this.id,
    required this.name,
    this.description,
    required this.bounds,
    required this.center,
    required this.defaultZoom,
    required this.minZoom,
    required this.maxZoom,
    required this.areas,
    this.trail,
  });

  factory Crag.fromJson(JsonObject json) {
    if (json
        case {
          'id': int id,
          'name': String name,
          'description': String? description,
          'bounds': JsonObject bounds,
          'center': JsonObject center,
          'defaultZoom': int defaultZoom,
          'minZoom': int minZoom,
          'maxZoom': int maxZoom,
          'areas': List<dynamic> areas,
          'trail': JsonObject? trail,
        }) {
      return Crag(
        id: id,
        name: name,
        description: description,
        bounds: Bounds.fromJson(bounds),
        center: Coordinate.fromJson(center),
        defaultZoom: defaultZoom,
        minZoom: minZoom,
        maxZoom: maxZoom,
        areas: areas.map((area) => Area.fromJson(area)).toList(),
        trail: trail != null ? Trail.fromJson(trail) : null,
      );
    }
    throw JSONException("Crag", json);
  }

  JsonObject toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'bounds': bounds.toJson(),
      'center': center.toJson(),
      'defaultZoom': defaultZoom,
      'minZoom': minZoom,
      'maxZoom': maxZoom,
      'areas': areas.map((area) => area.toJson()).toList(),
      'trail': trail?.toJson(),
    };
  }
}
